File: web/src/api/endpoints/upstream.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import type {
    NewAPITokenUsage,
    UpstreamAuthMode,
    UpstreamGatewayKey,
    UpstreamGroup,
    UpstreamInspectRequest,
    UpstreamInspectResult,
    UpstreamProviderType,
    UpstreamSubscription,
} from './channel';
import type { UpstreamModelPrice } from './model';

export type UpstreamCheckinLogEntry = {
    success: boolean;
    amount?: number;
    message?: string;
    at: string;
};

export type UpstreamSite = {
    id: number;
    name: string;
    provider_type: UpstreamProviderType;
    base_url: string;
    api_base_url: string;
    auth_mode: UpstreamAuthMode;
    enabled: boolean;
    auto_refresh: boolean;
    refresh_interval_secs: number;
    auto_checkin: boolean;
    checkin_interval_secs: number;
    last_checkin_at?: string;
    checkin_log?: string;
    sync_to_channel: boolean;
    linked_channel_id?: number;
    last_refresh_at?: string;
    last_refresh_status?: string;
    last_refresh_message?: string;
    model_count: number;
    key_count: number;
    group_count: number;
    price_count: number;
    subscription_count: number;
    balance_available: boolean;
    balance_used?: number;
    balance_remain?: number;
    balance_unlimited?: boolean;
    balance_alert_threshold?: number;
    last_balance_check_at?: string;
    last_balance_value?: number;
    auto_create_key?: boolean;
    key_quota_limit?: number;
    key_expire_days?: number;
    auto_sync_group?: boolean;
    auto_sync_price?: boolean;
    source_label?: string;
};

export type UpstreamCredential = {
    id: number;
    upstream_site_id: number;
    credential_type: string;
    auth_mode: UpstreamAuthMode;
    display_name: string;
    masked_value: string;
    user_id?: string;
    importable: boolean;
    last_validated_at?: string;
};

export type UpstreamKeySnapshot = {
    id: number;
    upstream_site_id: number;
    name?: string;
    masked_key?: string;
    allowed_models?: string;
    request_capabilities?: string;
    groups?: string;
    status?: string;
    quota?: number;
    quota_used?: number;
    expires_at?: string;
    importable: boolean;
    source_type?: string;
    channel_key_id?: number;
};

export type UpstreamGroupSnapshot = {
    id: number;
    upstream_site_id: number;
    external_id?: string;
    name: string;
    description?: string;
    platform?: string;
    status?: string;
    rate_multiplier?: number;
    models?: string;
    request_capabilities?: string;
    source?: string;
};

export type UpstreamSiteDetail = {
    site: UpstreamSite;
    credentials: UpstreamCredential[];
    keys: UpstreamKeySnapshot[];
    groups: UpstreamGroupSnapshot[];
    prices: UpstreamModelPrice[];
    subscriptions?: UpstreamSubscription[];
    linked_channel?: {
        id: number;
        name: string;
        type: number;
        enabled: boolean;
        key_count: number;
    };
};

export type UpstreamSiteCreateRequest = {
    name?: string;
    provider_type: UpstreamProviderType;
    base_url: string;
    auth_mode: UpstreamAuthMode;
    token?: string;
    access_key?: string;
    user_id?: string;
    username?: string;
    password?: string;
    auto_refresh?: boolean;
    refresh_interval_secs?: number;
    auto_checkin?: boolean;
    checkin_interval_secs?: number;
    sync_to_channel?: boolean;
    auto_sync_group?: boolean;
    auto_sync_price?: boolean;
    auto_create_key?: boolean;
    key_quota_limit?: number;
    key_expire_days?: number;
    channel_name?: string;
    target_channel_id?: number;
};

export type UpstreamSiteUpdateRequest = {
    id: number;
    name?: string;
    enabled?: boolean;
    auto_refresh?: boolean;
    refresh_interval_secs?: number;
    auto_checkin?: boolean;
    checkin_interval_secs?: number;
    sync_to_channel?: boolean;
    auto_sync_group?: boolean;
    auto_sync_price?: boolean;
    linked_channel_id?: number;
    balance_alert_threshold?: number;
    auto_create_key?: boolean;
    key_quota_limit?: number;
    key_expire_days?: number;
};

export type UpstreamRefreshRequest = {
    id: number;
    apply_channel?: boolean;
};

export type UpstreamApplyRequest = {
    id: number;
    target_channel_id?: number;
};

export type UpstreamCreateKeyRequest = {
    site_id: number;
    name: string;
    quota?: number;
    expires_at?: string;
    models?: string[];
    groups?: string[];
};

export type UpstreamCheckinResult = {
    success: boolean;
    amount?: number;
    message?: string;
    at: string;
};

export type UpstreamCreateKeyResult = {
    name: string;
    key: string;
    masked_key: string;
};

export type UpstreamHealthStatus = 'healthy' | 'degraded' | 'unhealthy';

export type UpstreamHealthItem = {
    id: number;
    name: string;
    status: UpstreamHealthStatus;
    enabled: boolean;
    last_refresh_at?: string;
    last_refresh_status?: string;
    model_count: number;
    balance_available: boolean;
    balance_remain: number;
    balance_unlimited: boolean;
    balance_alert: boolean;
    error_rate: number;
    suppressed: boolean;
    reasons?: string[];
};

export type UpstreamUsagePoint = {
    date: string;
    timestamp: number;
    request_count: number;
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
    cost: number;
    success_count: number;
    failure_count: number;
};

export type UpstreamUsageResponse = {
    site_id: number;
    days: number;
    points: UpstreamUsagePoint[];
    channel_ids?: number[];
};

export type UpstreamInspectPreview = UpstreamInspectResult & {
    keys?: UpstreamGatewayKey[];
    groups?: UpstreamGroup[];
    token_usage: NewAPITokenUsage;
};

export function useUpstreamSiteList() {
    return useQuery({
        queryKey: ['upstream', 'list'],
        queryFn: async () => apiClient.get<UpstreamSite[]>('/api/v1/upstream/list'),
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

export function useUpstreamSiteDetail(id?: number) {
    return useQuery({
        queryKey: ['upstream', 'detail', id ?? 'none'],
        queryFn: async () => apiClient.get<UpstreamSiteDetail>(`/api/v1/upstream/detail/${id}`),
        enabled: Boolean(id),
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

export function useInspectUpstreamSite() {
    return useMutation({
        mutationFn: async (data: UpstreamInspectRequest) => apiClient.post<UpstreamInspectPreview>('/api/v1/upstream/inspect', data),
    });
}

export function useCreateUpstreamSite() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: UpstreamSiteCreateRequest) => apiClient.post<UpstreamSiteDetail>('/api/v1/upstream/create', data),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
            queryClient.invalidateQueries({ queryKey: ['channels', 'list'] });
            queryClient.invalidateQueries({ queryKey: ['models'] });
        },
    });
}

export function useUpdateUpstreamSite() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: UpstreamSiteUpdateRequest) => apiClient.post<UpstreamSite>('/api/v1/upstream/update', data),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
        },
    });
}

export function useRefreshUpstreamSite() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: UpstreamRefreshRequest) => apiClient.post<UpstreamSiteDetail>('/api/v1/upstream/refresh', data),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
            queryClient.invalidateQueries({ queryKey: ['channels', 'list'] });
            queryClient.invalidateQueries({ queryKey: ['models'] });
        },
    });
}

export function useCheckinUpstreamSite() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (id: number) => apiClient.post<UpstreamCheckinResult>('/api/v1/upstream/checkin', { id }),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
        },
    });
}

export function useApplyUpstreamSite() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: UpstreamApplyRequest) => apiClient.post('/api/v1/upstream/apply', data),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
            queryClient.invalidateQueries({ queryKey: ['channels', 'list'] });
            queryClient.invalidateQueries({ queryKey: ['models'] });
        },
    });
}

export function useDeleteUpstreamSite() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (id: number) => apiClient.delete<null>(`/api/v1/upstream/delete/${id}`),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
            queryClient.invalidateQueries({ queryKey: ['models'] });
        },
    });
}

export function useCreateUpstreamKey() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: UpstreamCreateKeyRequest) => apiClient.post<UpstreamCreateKeyResult>('/api/v1/upstream/create-key', data),
        onSuccess: (_data, variables) => {
            queryClient.invalidateQueries({ queryKey: ['upstream', 'detail', String(variables.site_id)] });
            queryClient.invalidateQueries({ queryKey: ['upstream', 'list'] });
        },
    });
}

export function useUpstreamSiteHealth() {
    return useQuery({
        queryKey: ['upstream', 'health'],
        queryFn: async () => apiClient.get<UpstreamHealthItem[]>('/api/v1/upstream/health'),
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

export function useUpstreamSiteUsage(id?: number, days = 7) {
    return useQuery({
        queryKey: ['upstream', 'usage', id ?? 'none', days],
        queryFn: async () => apiClient.get<UpstreamUsageResponse>(`/api/v1/upstream/usage/${id}?days=${days}`),
        enabled: Boolean(id),
        refetchInterval: 60000,
        refetchOnMount: 'always',
    });
}

export function useRestoreUpstreamPriority() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (id: number) => apiClient.post<null>(`/api/v1/upstream/restore-priority/${id}`),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['upstream'] });
        },
    });
}

File: internal/model/upstream.go
package model

import "time"

const (
	UpstreamRefreshManual    = "manual"
	UpstreamRefreshScheduled = "scheduled"

	UpstreamCredentialManagementToken = "management_token"
	UpstreamCredentialAccessKey       = "access_key"

	UpstreamPriceSourceGateway = "gateway"
)

type UpstreamSite struct {
	ID                    int       `json:"id" gorm:"primaryKey"`
	Name                  string    `json:"name" gorm:"not null;index"`
	ProviderType          string    `json:"provider_type" gorm:"index"`
	BaseURL               string    `json:"base_url"`
	APIBaseURL            string    `json:"api_base_url"`
	AuthMode              string    `json:"auth_mode"`
	Enabled               bool      `json:"enabled" gorm:"default:true"`
	AutoRefresh           bool      `json:"auto_refresh" gorm:"default:false"`
	RefreshIntervalSecs   int       `json:"refresh_interval_secs" gorm:"default:43200"`
	AutoCheckin           bool      `json:"auto_checkin" gorm:"default:false"`
	CheckinIntervalSecs   int       `json:"checkin_interval_secs" gorm:"default:86400"`
	LastCheckinAt         time.Time `json:"last_checkin_at,omitempty"`
	CheckinLog            string    `json:"checkin_log,omitempty" gorm:"type:text"`
	SyncToChannel         bool      `json:"sync_to_channel" gorm:"default:false"`
	LinkedChannelID       int       `json:"linked_channel_id" gorm:"index"`
	LastRefreshAt         time.Time `json:"last_refresh_at,omitempty"`
	LastRefreshStatus     string    `json:"last_refresh_status,omitempty"`
	LastRefreshMessage    string    `json:"last_refresh_message,omitempty" gorm:"type:text"`
	ModelCount            int       `json:"model_count"`
	KeyCount              int       `json:"key_count"`
	GroupCount            int       `json:"group_count"`
	PriceCount            int       `json:"price_count"`
	SubscriptionCount     int       `json:"subscription_count"`
	BalanceAvailable      bool      `json:"balance_available"`
	BalanceUsed           float64   `json:"balance_used"`
	BalanceRemain         float64   `json:"balance_remain"`
	BalanceUnlimited      bool      `json:"balance_unlimited"`
	BalanceAlertThreshold float64   `json:"balance_alert_threshold" gorm:"default:0"`
	LastBalanceCheckAt    time.Time `json:"last_balance_check_at,omitempty"`
	LastBalanceValue      float64   `json:"last_balance_value"`
	AutoCreateKey         bool      `json:"auto_create_key" gorm:"default:false"`
	KeyQuotaLimit         float64   `json:"key_quota_limit" gorm:"default:0"`
	KeyExpireDays         int       `json:"key_expire_days" gorm:"default:0"`
	AutoSyncGroup         bool      `json:"auto_sync_group" gorm:"default:false"`
	AutoSyncPrice         bool      `json:"auto_sync_price" gorm:"default:false"`
	SourceLabel           string    `json:"source_label,omitempty"`
	CreatedAt             time.Time `json:"created_at"`
	UpdatedAt             time.Time `json:"updated_at"`
}

type UpstreamCheckinLogEntry struct {
	Success bool      `json:"success"`
	Amount  float64   `json:"amount,omitempty"`
	Message string    `json:"message,omitempty"`
	At      time.Time `json:"at"`
}

type UpstreamCheckinResult struct {
	Success bool      `json:"success"`
	Amount  float64   `json:"amount,omitempty"`
	Message string    `json:"message,omitempty"`
	At      time.Time `json:"at"`
}

type UpstreamCredential struct {
	ID              int       `json:"id" gorm:"primaryKey"`
	UpstreamSiteID  int       `json:"upstream_site_id" gorm:"index"`
	CredentialType  string    `json:"credential_type" gorm:"index"`
	AuthMode        string    `json:"auth_mode"`
	DisplayName     string    `json:"display_name"`
	MaskedValue     string    `json:"masked_value"`
	EncryptedValue  string    `json:"-" gorm:"type:text"`
	UserID          string    `json:"user_id,omitempty"`
	Importable      bool      `json:"importable"`
	LastValidatedAt time.Time `json:"last_validated_at,omitempty"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

type UpstreamKeySnapshot struct {
	ID                  int       `json:"id" gorm:"primaryKey"`
	UpstreamSiteID      int       `json:"upstream_site_id" gorm:"index"`
	Name                string    `json:"name,omitempty"`
	MaskedKey           string    `json:"masked_key,omitempty"`
	AllowedModels       string    `json:"allowed_models,omitempty" gorm:"type:text"`
	RequestCapabilities string    `json:"request_capabilities,omitempty"`
	Groups              string    `json:"groups,omitempty" gorm:"type:text"`
	Status              string    `json:"status,omitempty"`
	Quota               float64   `json:"quota,omitempty"`
	QuotaUsed           float64   `json:"quota_used,omitempty"`
	ExpiresAt           string    `json:"expires_at,omitempty"`
	Importable          bool      `json:"importable"`
	SourceType          string    `json:"source_type,omitempty"`
	ChannelKeyID        int       `json:"channel_key_id,omitempty" gorm:"index"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
}

type UpstreamGroupSnapshot struct {
	ID                  int       `json:"id" gorm:"primaryKey"`
	UpstreamSiteID      int       `json:"upstream_site_id" gorm:"index"`
	ExternalID          string    `json:"external_id,omitempty"`
	Name                string    `json:"name"`
	Description         string    `json:"description,omitempty" gorm:"type:text"`
	Platform            string    `json:"platform,omitempty"`
	Status              string    `json:"status,omitempty"`
	RateMultiplier      float64   `json:"rate_multiplier,omitempty"`
	Models              string    `json:"models,omitempty" gorm:"type:text"`
	RequestCapabilities string    `json:"request_capabilities,omitempty"`
	Source              string    `json:"source,omitempty"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
}

type UpstreamModelPrice struct {
	ID                 int       `json:"id" gorm:"primaryKey"`
	UpstreamSiteID     int       `json:"upstream_site_id" gorm:"index"`
	ChannelID          int       `json:"channel_id,omitempty" gorm:"index"`
	ModelName          string    `json:"model_name" gorm:"index"`
	CanonicalName      string    `json:"canonical_name,omitempty" gorm:"index"`
	PriceSource        string    `json:"price_source,omitempty"`
	PriceMatchedKey    string    `json:"price_matched_key,omitempty"`
	SourceLabel        string    `json:"source_label,omitempty"`
	CachePolicy        string    `json:"cache_policy,omitempty"`
	CacheReason        string    `json:"cache_reason,omitempty" gorm:"type:text"`
	CacheSupported     *bool     `json:"cache_supported,omitempty" gorm:"-"`
	Input              float64   `json:"input"`
	Output             float64   `json:"output"`
	CacheRead          float64   `json:"cache_read"`
	CacheWrite         float64   `json:"cache_write"`
	OfficialInput      float64   `json:"official_input"`
	OfficialOutput     float64   `json:"official_output"`
	OfficialCacheRead  float64   `json:"official_cache_read"`
	OfficialCacheWrite float64   `json:"official_cache_write"`
	UpdatedAt          time.Time `json:"updated_at"`
	CreatedAt          time.Time `json:"created_at"`
}

type UpstreamSiteCreateRequest struct {
	Name                string  `json:"name"`
	ProviderType        string  `json:"provider_type"`
	BaseURL             string  `json:"base_url"`
	AuthMode            string  `json:"auth_mode"`
	Token               string  `json:"token,omitempty"`
	AccessKey           string  `json:"access_key,omitempty"`
	UserID              string  `json:"user_id,omitempty"`
	Username            string  `json:"username,omitempty"`
	Password            string  `json:"password,omitempty"`
	AutoRefresh         bool    `json:"auto_refresh"`
	RefreshIntervalSecs int     `json:"refresh_interval_secs"`
	AutoCheckin         bool    `json:"auto_checkin"`
	CheckinIntervalSecs int     `json:"checkin_interval_secs"`
	SyncToChannel       bool    `json:"sync_to_channel"`
	AutoSyncGroup       bool    `json:"auto_sync_group"`
	AutoSyncPrice       bool    `json:"auto_sync_price"`
	AutoCreateKey       bool    `json:"auto_create_key"`
	KeyQuotaLimit       float64 `json:"key_quota_limit"`
	KeyExpireDays       int     `json:"key_expire_days"`
	ChannelName         string  `json:"channel_name,omitempty"`
	TargetChannelID     int     `json:"target_channel_id,omitempty"`
}

type UpstreamSiteUpdateRequest struct {
	ID                    int      `json:"id" binding:"required"`
	Name                  *string  `json:"name,omitempty"`
	Enabled               *bool    `json:"enabled,omitempty"`
	AutoRefresh           *bool    `json:"auto_refresh,omitempty"`
	RefreshIntervalSecs   *int     `json:"refresh_interval_secs,omitempty"`
	AutoCheckin           *bool    `json:"auto_checkin,omitempty"`
	CheckinIntervalSecs   *int     `json:"checkin_interval_secs,omitempty"`
	SyncToChannel         *bool    `json:"sync_to_channel,omitempty"`
	AutoSyncGroup         *bool    `json:"auto_sync_group,omitempty"`
	AutoSyncPrice         *bool    `json:"auto_sync_price,omitempty"`
	LinkedChannelID       *int     `json:"linked_channel_id,omitempty"`
	BalanceAlertThreshold *float64 `json:"balance_alert_threshold,omitempty"`
	AutoCreateKey         *bool    `json:"auto_create_key,omitempty"`
	KeyQuotaLimit         *float64 `json:"key_quota_limit,omitempty"`
	KeyExpireDays         *int     `json:"key_expire_days,omitempty"`
}

type UpstreamBalanceCheckResult struct {
	Alert     bool    `json:"alert"`
	Threshold float64 `json:"threshold"`
	Remain    float64 `json:"remain"`
	Message   string  `json:"message,omitempty"`
}

type UpstreamSiteDetail struct {
	Site          UpstreamSite            `json:"site"`
	Credentials   []UpstreamCredential    `json:"credentials"`
	Keys          []UpstreamKeySnapshot   `json:"keys"`
	Groups        []UpstreamGroupSnapshot `json:"groups"`
	Prices        []UpstreamModelPrice    `json:"prices"`
	Subscriptions []UpstreamSubscription  `json:"subscriptions,omitempty"`
	LinkedChannel *UpstreamAppliedChannel `json:"linked_channel,omitempty"`
}

type UpstreamRefreshRequest struct {
	ID           int  `json:"id" binding:"required"`
	ApplyChannel bool `json:"apply_channel"`
	Manual       bool `json:"manual"`
}

type UpstreamCreateKeyRequest struct {
	SiteID    int      `json:"site_id" binding:"required"`
	Name      string   `json:"name" binding:"required"`
	Quota     float64  `json:"quota"`
	ExpiresAt string   `json:"expires_at,omitempty"`
	Models    []string `json:"models,omitempty"`
	Groups    []string `json:"groups,omitempty"`
}

type UpstreamCreateKeyResult struct {
	Name      string `json:"name"`
	Key       string `json:"key"`
	MaskedKey string `json:"masked_key"`
}

type UpstreamPriceSummary struct {
	ModelName        string               `json:"model_name"`
	OfficialPrice    OfficialLLMPrice     `json:"official_price"`
	GatewayPrices    []UpstreamModelPrice `json:"gateway_prices"`
	EffectiveGateway *UpstreamModelPrice  `json:"effective_gateway,omitempty"`
}

type UpstreamHealthStatus string

const (
	UpstreamHealthStatusHealthy   UpstreamHealthStatus = "healthy"
	UpstreamHealthStatusDegraded  UpstreamHealthStatus = "degraded"
	UpstreamHealthStatusUnhealthy UpstreamHealthStatus = "unhealthy"
)

type UpstreamHealthItem struct {
	ID                int                  `json:"id"`
	Name              string               `json:"name"`
	Status            UpstreamHealthStatus `json:"status"`
	Enabled           bool                 `json:"enabled"`
	LastRefreshAt     time.Time            `json:"last_refresh_at,omitempty"`
	LastRefreshStatus string               `json:"last_refresh_status,omitempty"`
	ModelCount        int                  `json:"model_count"`
	BalanceAvailable  bool                 `json:"balance_available"`
	BalanceRemain     float64              `json:"balance_remain"`
	BalanceUnlimited  bool                 `json:"balance_unlimited"`
	BalanceAlert      bool                 `json:"balance_alert"`
	ErrorRate         float64              `json:"error_rate"`
	Suppressed        bool                 `json:"suppressed"`
	Reasons           []string             `json:"reasons,omitempty"`
}

type UpstreamUsagePoint struct {
	Date         string  `json:"date"`
	Timestamp    int64   `json:"timestamp"`
	RequestCount int64   `json:"request_count"`
	InputTokens  int64   `json:"input_tokens"`
	OutputTokens int64   `json:"output_tokens"`
	TotalTokens  int64   `json:"total_tokens"`
	Cost         float64 `json:"cost"`
	SuccessCount int64   `json:"success_count"`
	FailureCount int64   `json:"failure_count"`
}

type UpstreamUsageResponse struct {
	SiteID     int                  `json:"site_id"`
	Days       int                  `json:"days"`
	Points     []UpstreamUsagePoint `json:"points"`
	ChannelIDs []int                `json:"channel_ids,omitempty"`
}
